### ISSUE TITLE
how to run the largest possible model on a single A100 80Gb

### ISSUE THREAD

**Issue by huey2531** [facebookresearch/llama#101]
I was able to get the 7B model to work. It looks like I might be able to run the 33B version?
Will I need to merge the checkpoint files (.pth) to run on a single GPU? and set MP = 1?

It would be great if FAIR could provide some guidance on vram requirements

**Comment by pauldog** [comment]
I had an idea that you might use GPU virtualisation to make it appear like you had two GPUs. Well, it wasn't my idea, it was Chat GPTs.

**Comment by cedrickchee** [comment]
> It would be great if FAIR could provide some guidance on vram requirements

See: [memory requirements for each model size](https://github.com/cedrickchee/llama/blob/main/chattyllama/hardware.md#memory-requirements-for-each-model-size).

It's a small place that's part of my own experiments where I also document down things I learned. While the official documentations are lacking now, you can also learn from the good discussions around this project on various GitHub Issues.

If I'm not wrong, 65B needs a GPU cluster with a total of 135GB VRAM.

(I'm not affiliated with FAIR.)

**Comment by benob** [comment]
> I had an idea that you might use GPU virtualisation to make it appear like you had two GPUs. Well, it wasn't my idea, it was Chat GPTs.

Seems that torch.distributed does not support GPU virtualization.

**Comment by benob** [comment]
I was able to run the 13B and 30B models on a single A100-80GB. I used the following script to reshard the models: https://gist.github.com/benob/4850a0210b01672175942203aa36d300

**Comment by baleksey** [comment]
@benob Thanks for the script! I was successfully run 13B with it. But I was failed while sharding 30B model as I run our of memory (128 RAM is obviously not enough for this). Is it any way you can share your combined 30B model so I can try to run it on my A6000-48GB? Thank you so much in advanced!

**Comment by USBhost** [comment]
> @benob Thanks for the script! I was successfully run 13B with it. But I was failed while sharding 30B model as I run our of memory (128 RAM is obviously not enough for this). Is it any way you can share your combined 30B model so I can try to run it on my A6000-48GB? Thank you so much in advance!

For one can't even run the 33B model in 16bit mod. You would need another 16GB+ of vram. 

But using the 8bit fork you can load it without needing to reshard it.

**Comment by baleksey** [comment]
@USBhost Got it, thanks for clarification! Can you direct me to 8bit model/code for 33B l so I can continue my experiments? :)

**Comment by USBhost** [comment]
> @USBhost Got it, thanks for clarification! Can you direct me to 8bit model/code for 33B l so I can continue my experiments? :)

https://github.com/oobabooga/text-generation-webui/issues/147#issuecomment-1454850191

I'm also a fellow A6000 user.

**Comment by baleksey** [comment]
@USBhost Thank you! Have you managed to run 33B model with it? I still have OOMs after model quantization..

**Comment by USBhost** [comment]
> @USBhost Thank you! Have you managed to run 33B model with it? I still have OOMs after model quantization..

I'm using ooba python server.py --listen --model LLaMA-30B --load-in-8bit --cai-chat

If you want just to use LLaMA-8bit then only run with node 1.

**Comment by MarkSchmidty** [comment]
> > It would be great if FAIR could provide some guidance on vram requirements
> 
> See: [memory requirements for each model size](https://github.com/cedrickchee/llama/blob/main/chattyllama/hardware.md#memory-requirements-for-each-model-size).
> 
> It's a small place that's part of my own experiments where I also document down things I learned. While the official documentations are lacking now, you can also learn from the good discussions around this project on various GitHub Issues.
> 
> If I'm not wrong, 65B needs a GPU cluster with a total of 250GB in fp16 precision or half in int8.
> 
> (I'm not affiliated with FAIR.)

The linked memory requirement calculation table is wrong, I think. The corrected table should look like:

Memory requirements in 8-bit precision:
| Model (on disk)***       | 13   | 24   | 60   | 120  |
|--------------------------|------|------|------|------|
| Memory Requirements (GB) | 6.7  | 13   | 32.5 | 65.2 |
| Cache                    | 1    | 2    | 3    | 5    |
|                          |      |      |      |      |
| Total                    | 7.7   | 16   | 37.5   | 70.2  |

This seems to more closely match up with what I'm seeing people report their actual VRAM usage is in https://github.com/oobabooga/text-generation-webui/issues/147

For example:
LLaMA-13B: 16249MiB
LLaMA-7B: 9225MiB

If this is true then 65B should fit on a single A100 80GB after all.

**Comment by cedrickchee** [comment]
@MarkSchmidty

> The linked memory requirement calculation table is adding the wrong rows together
Not exactly.

> The corrected table should look like: ...

Yours looks almost correct. I have updated the table based on your findings (a copy below).

> Memory requirements in 8-bit precision:

To prevent all sort of confusion, let's keep the precision in fp16 (before 8-bit quantization).

I need to point out that when people report their actual VRAM, they never state the model arguments. The most important ones are `max_batch_size` and `max_seq_length`. These impact the VRAM required (too large, you run into OOM. FAIR should really set the `max_batch_size` to 1 by default. It's 32 now.)

Based on the [Transformer kv cache formula](https://kipp.ly/blog/transformer-inference-arithmetic/#capacity)
and `max_batch_size` of 1 and `max_seq_length` of 1024, the table looks like this now:

Memory requirements in fp16 precision.

| Model Params (Billions)  | 6.7  | 13   | 32.5 | 65.2 |
|--------------------------|------|------|------|------|
| Model on disk (GB)***    | 13   | 26   | 65   | 130  |
| Cache (GB)               | 1    | 1    | 2    | 3    |
| Total (GB)               | 14   | 27   | 67   | 133  |

> This seems to more closely match up with what I'm seeing people report their actual VRAM usage

Sounds right to me. I have also check the above model mem. requirements against GPU requirements ([from my repo](https://github.com/cedrickchee/llama/blob/main/chattyllama/hardware.md#guide-on-gpu-requirements)). The two closely match up now.

> If this is true then 65B should fit on a single A100 80GB after all.

133GB in fp16, 66.5GB in int8. 66.5GB VRAM is within the range of 1x A100 80GB. So, should fit but can't confirm though.

**Comment by zhongtao93** [comment]
I found download failed for 65B params.json, anyone can kindly provide it.
Thank you very much!!!
<img width="202" alt="image" src="https://user-images.githubusercontent.com/15263824/223388639-d8d26be6-ede2-4bd4-a436-1160aed2943b.png">

**Comment by MarkSchmidty** [comment]
4-bit for LLaMA is underway https://github.com/oobabooga/text-generation-webui/issues/177

65B in int4 fits on a single v100 40GB, even further reducing the cost to access this powerful model.

**Comment by cedrickchee** [comment]
> 4-bit for LLaMA

I'm well aware. It's a matter of time.
This 4-bit post-training quantization technique is different the one that Tim Dettmers working on.

> 65B in int4 fits on a single v100 40GB, even further reducing the cost to access this powerful model.

Exciting but the big question is, does this technique reduce the model performance (both inference throughput and accuracy)? To my understanding, I couldn't find legible results that prove the model performance closely match LLM.int8().

> Int4 LLaMA VRAM usage is aprox. as follows:

Thanks for the heads-up. I will update and add this to my repo accordingly.

**Comment by MarkSchmidty** [comment]
That particular 4-bit implementation is mostly a proof of concept at this point. Bitsandbytes may be getting some 4-bit functionality towards the end of this month. Best to wait for that.

**Comment by ayeeyecorp** [comment]
> I was able to run the 13B and 30B (batch size 1) models on a single A100-80GB. I used a script [1] to reshard the models and torchrun with --nproc_per_node 1.
> 
> [1] https://gist.github.com/benob/4850a0210b01672175942203aa36d300

So I was able to reshard on a Lambda a6000 instance and the 13B single shard file worked great and inference was successful using example.py from the repo. 

The checksum (sha256sum) of the new consolidated.00.pth and params.json files yielded:
3990e677b13fb3afc029e131b623f609fb7dce2afb25cc16b757e170e4633c60  consolidated.00.pth
4ab77bec4d4405ccb66a97b282574c89a94417e3c32e5f68f37e2876fc21322f  params.json

However, when I reshard on my local PC with that script, I cannot for the life of me get the same checksums to match. I reinstalled Ubuntu to make sure there was not some pre-installed package that was corrupting the reshard.py script. I am running the same NVIDIA driver and am (hopefully) using Pytorch to match what is running on the Lambda A6000 instance. I say hopefully, because my torch.__version__ yields: 

1.12.1+cu116

but the lambda instance only states:

1.12.1

Could this be causing the issue?

This command I used to install pytroch (which seems to be the only relevant dependency) for the reshard.py script is:
pip install torch==1.12.1+cu116 torchvision==0.13.1+cu116 torchaudio==0.12.1 --extra-index-url https://download.pytorch.org/whl/cu116

Thoughts? Everything is installed in a pyenv and now I am totally stuck. Thanks for the help. @benob
